### ISSUE TITLE
yapf on save not working

### ISSUE THREAD

**Issue by pranavsubramani** [syl20bnr/spacemacs#9766]
I used to use emacs (not a power user or anything, just kinda liked it) and recently switched to spacemacs, I love the python layer and everything from the autocompletion to the linter works. The only thing that doesn't work is the autopep8 on save.

Example:
`a = [1,2,3]`

Now, pep8 requires this to be `[1, 2, 3]`

But for some reason when I `C-x C-s`, it doesn't automatically do it.

I set this in my dotspacemacs file

`(setq-default dotspacemacs-configuration-layers '(
  (python :variables python-enable-yapf-format-on-save t)))`

I have also installed yapf

Sorry if this is a noobish question, just couldn't find resources online.

**Comment by ellakk** [comment]
Works fine for me. Are you sure yapf are available in your path? If you are: 
1. Open a terminal. Create `test.py` and `add a = [1,2,3]`
2. Run `yapf ./test.py` in your terminal in the same directory as the file.
2.1. If output of yapf is command not found then yapf is not in your path.
2.2. if output from yapf is `a = [1,2,3]` then something is wrong with your yapf configuration.
2.3 if output from yapf is `a = [1, 2, 3]` then the probable cause is that your environment variables is not set correctly for emacs, check 3.
3. Open spacemacs and run eshell (M-x eshell). Check if yapf is available in eshell with command `which yapf`, if it's not, add the path where yapf is located to your [exec-path](https://www.emacswiki.org/emacs/ExecPath) in emacs.

Also what os are you running?

**Comment by pranavsubramani** [comment]
Hey @ellakk , I tried the yapf on the test.py and I got `a = [1, 2, 3]`

I opened M-x eshell and typed which yapf and it says

/home/pranav/.local/bin/yapf (which is the same one I get from my terminal)

I'm running Ubuntu 16.04

**Comment by ellakk** [comment]
Seems on master there is a bug on the python layer on master that makes sure the yapf-on-save hook never gets added, its fixed on develop. You can change to develop branch or just add the hook your self to your `dotspacemacs/user-config` block like this:
`(add-hook 'python-mode-hook 'yapf-mode)`

**Comment by ellakk** [comment]
That's odd. Does it work if you just do command M-x`yapfify-buffer` in the buffer?

**Comment by pranavsubramani** [comment]
Yeah that works :/ But again, it doesn't work when I use `C-x C-s` like it used to in emacs

Could it be the position of the `(setq-default dotspacemacs-configuration-layers '( (python :variables python-enable-yapf-format-on-save t)))`? (It's both inside the user config and in global scope)

**Comment by ellakk** [comment]
I'm not sure i understand, you should only have one `dotspacemacs-configuration-layers` in your init file and it should be inside the defun `dotspacemacs/layers` where it is in the official template, like this:
```emacs-lisp
(defun dotspacemacs/layers ()
  (setq-default
   dotspacemacs-distribution 'spacemacs
   dotspacemacs-enable-lazy-installation 'unused
   dotspacemacs-ask-for-lazy-installation t
   dotspacemacs-configuration-layer-path '()
   dotspacemacs-configuration-layers
   '(
     helm
     emacs-lisp
     (python :variables python-enable-yapf-format-on-save t)
     )
   dotspacemacs-additional-packages '()
   dotspacemacs-frozen-packages '()
   dotspacemacs-excluded-packages '()
   dotspacemacs-install-packages 'used-only))
```

**Comment by pranavsubramani** [comment]
Hey, I'm so sorry, I'm really new to spacemacs and doing that fixed it for me, I also used this issue which worked out fairly well, thanks again! (https://github.com/syl20bnr/spacemacs/issues/8159)

**Comment by ellakk** [comment]
No problems, I'm also a new Spacemacs user but I've used Emacs for about 12 years.
Also that issue you referred to was likely fixed by #7804 but they have just forgotten to close it.

**Comment by pranavsubramani** [comment]
Ah sounds awesome, I've only been coding for about 3-5 years now (primarily used sublime text before). I'm going to go ahead and close this :+1:
